### ISSUE TITLE
Allow delete of un-mounted/un-mountable pool

### ISSUE THREAD

**Issue by phillxnet** [rockstor/rockstor-core#2395]
Thanks to forum member qubes for helping to highlight this usability short-fall. In certain instances such as when a non redundant pool has a missing member, the recovery scenario is to delete the pool and start over from backups; see:

"Data Recovery: btrfs-raid0 and btrfs-single Pools":
https://rockstor.com/docs/data_loss.html#data-recovery-btrfs-raid0-and-btrfs-single-pools

However we currently fail on the Web-UI side by paradoxically dis-allowing this action due to the pool not being mounted initially!!

>  Error running a command. cmd = /usr/bin/umount -l /mnt2/test-pool-balance. rc = 32. stdout = ['']. stderr = ['umount: /mnt2/test-pool-balance: not mounted.', ''] 

>             Traceback (most recent call last):
>   File "/opt/rockstor-dev/src/rockstor/rest_framework_custom/generic_view.py", line 41, in _handle_exception
>     yield
>   File "/opt/rockstor-dev/src/rockstor/storageadmin/views/pool.py", line 800, in delete
>     umount_root(pool_path)
>   File "/opt/rockstor-dev/src/rockstor/fs/btrfs.py", line 600, in umount_root
>     raise ce
> CommandException: Error running a command. cmd = /usr/bin/umount -l /mnt2/test-pool-balance. rc = 32. stdout = ['']. stderr = ['umount: /mnt2/test-pool-balance: not mounted.', '']

This is a poor user experience and we already warn of all data shares etc being removed and should thus be able to proceed as a user might expect and allow for the pool delete in this case. To avoid the paradoxical mount requirement prior to pool delete (unmount). The reasoning here is to first wipe all contents but for a broken pool that is obviously not possible.

**Comment by phillxnet** [comment]
@mlincett
Re:
> Can this be solved with a simple `if`? For example:

Hopefully! But I think it will need a little more alteration in our sanity checks, as they are just a little not-sane obviously. When last I looked it was just a matter of a corner case that we just don't deal with. So likely we just have to allow for database entry removals without pool wipe confirmation (which is why we fail as we check for a mount before proceeding with the pool wipe). So it would be good to log this less usual action accordingly so there is a record of a db-only action without fs confirmation.

Have you tried any modifications in a development install/instance yourself? This issue is actually earmarked to be addressed before our next stable release. So your input here is quite timely as it goes. Note however that we have now, in the testing branch (just tagged), moved to a way more straight-forward build method via Poetry. But for a python change such as this one just needs to restart the relevant rockstor systemd service (or all of them just in case). Or just reboot.

**Comment by mlincett** [comment]
> > Can this be solved with a simple `if`? For example:
> 
> Hopefully! But I think it will need a little more alteration in our sanity checks, as they are just a little not-sane obviously. When last I looked it was just a matter of a corner case that we just don't deal with. So likely we just have to allow for database entry removals without pool wipe confirmation (which is why we fail as we check for a mount before proceeding with the pool wipe). So it would be good to log this less usual action accordingly so there is a record of a db-only action without fs confirmation.
> 
> Have you tried any modifications in a development install/instance yourself? This issue is actually earmarked to be addressed before our next stable release. So your input here is quite timely as it goes. Note however that we have now, in the testing branch (just tagged), moved to a way more straight-forward build method via Poetry. But for a python change such as this one just needs to restart the relevant rockstor systemd service (or all of them just in case). Or just reboot.

Thank you for the explanation, I am in the process of setting up a development instance, I am happy to try the current testing branch there.

**Comment by phillxnet** [comment]
> Thank you for the explanation, I am in the process of setting up a development instance, I am happy to try the current testing branch there.

That good. All our development currently takes place only in testing as we have just changed how we build. Just take look at the contents of build.sh in latest testing as we have yet to update our docs on how things are to be build from now onwards. The latest testing channel 4.5.4-0 rpm (just released) may be a quick route to taking a look. It can, in turn, also be used to experiment on as all code and tests are also included. But for actual contribution you will still need to take the git approach. 

In short build.sh now does all that is required to take source to running instance, but you must locate the code (for the time being at least) in the rpm default location of /opt/rockstor (our old build system allowed for differences here).

**Comment by mlincett** [comment]
I was about to perform a test, but it turns out that removing all the (detached) devices from the (unmounted) pool results in its deletion so this seems a viable workaround (temporarily).

**Comment by phillxnet** [comment]
@mlincett Hello again.
Yes, re:
> it turns out that removing all the (detached) devices from the (unmounted) pool results in its deletion

That is by design as I remember and was an improvement we made a little while back, but this issue covers a slightly different circumstance.

For the wipe pool if all detached devices see:

For the circumstance you mention see:
"failure to remove detached disks via pool resize. Fixes #2099" #2184 https://github.com/rockstor/rockstor-core/pull/2184
There the testing section has the following:

> "Removing a single detached member from pools that have only detached members results, by design, in all detached members having their pool association removed, and on next pool state refresh, the pool itself being removed. See #2077 detailed caveat below."

There are a surprising number of corner cases with disk and pool management as it turns out !! That one above plagued us for quite a while and I'm quite pleased with the progress made at that time.

The above was included in our 3.9.2-60 release: 
https://github.com/rockstor/rockstor-core/releases/tag/3.9.2-60
Mid 2020.

Taking a look at the above pull request may help with locating the relevant code, if you haven't already. But there have been a number of improvements on from then and we now use Huey to schedule the longer running tasks such as disk removal. Adding disks is trivial and takes not time, except for the explicit balance we do there-after which we also schedule and thread via Huey. But when one removes a disk the btrfs subsystem does it's own internal balance that does not show up in a balance status command. This is currently a poblem of sorts for us and we have to infer this situation via our own analysis of btrfs fi usage I think it is. But all good and currently working for us, we just don't have percentage capability that way. But hopefully in time this internal balance will be surfaced by the btrfs tools more explicitly.

**Comment by phillxnet** [comment]
I am currently working on this issue.

**Comment by phillxnet** [comment]
It turns out that the core (but not all) of this issue is down to the following original text detailed exception:

`CommandException: Error running a command. cmd = /usr/bin/umount -l /mnt2/scratch-pool. rc = 32. stdout = ['']. stderr = ['umount: /mnt2/scratch-pool: not mounted.', '']`

Which we intent to catch but fail to do so due to the full stop !!

https://github.com/rockstor/rockstor-core/blob/e5c3126dc11a5bdbc993dbc3988902ae10e55227/src/rockstor/fs/btrfs.py#L808-L809

it should read:

```
                if re.search("not mounted\.$", l) is not None:
```
Which then allows for us to successfully catch this exception while honouring all others.

I am now preparing a pull request with this fix and some associated improvements in this area.

**Comment by phillxnet** [comment]
Closing as:
Fixed by #2560
